packages/iota/src/validation/outputs/common.ts

```diff
@@ -0,0 +1,277 @@
+    validateAmountIsLesserThanMaxSupply(output.amount, protocolInfo.tokenSupply, outputName);
+    validateNativeTokens(output.nativeTokens);
+
+    if (output.unlockConditions) {
+        const min = outputTypeValidations.minUnlockConditions;
+        const max = outputTypeValidations.unlockConditions.length;
+
+        validateCount(output.unlockConditions.length, min, max, `${outputName} Unlock Conditions`);
+        validateUnlockConditionAllowedTypes(outputType, output.unlockConditions, outputName);
+    }
+
+    if (output.features) {
+        const max = outputTypeValidations.features.length;
+        validateCount(output.features.length, 0, max, `${outputName} Features`);
+        validateFeatureAllowedTypes(outputType, output.features, outputName);
+    }
+
+    if (outputType !== BASIC_OUTPUT_TYPE && output.immutableFeatures && outputTypeValidations.immutableFeatures) {
+        const max = outputTypeValidations.immutableFeatures.length;
+
+        validateCount(output.immutableFeatures.length, 0, max, `${outputName} Immutable Features`);
+        validateImmutableFeatureAllowedTypes(outputType, output.immutableFeatures, outputName);
+    }
+}
+
+/**
+ * Validate the amount is greater than zero.
+ * @param amount The amount to validate.
+ * @param outputName The name of the output to use in the error message.
+ * @throws Error if the validation fails.
+ */
+function validateAmountIsGreaterThanZero(amount: string, outputName: string) {
+    if (bigInt(amount).leq(bigInt.zero)) {
+        failValidation(`${outputName} amount field must be greater than zero.`);
+    }
+}
+
+/**
+ * Validate the amount is lesser than maximum token supply.
+ * @param amount The amount to validate.
+ * @param tokenSupply The tokken supply amount.
+ * @param outputName The name of the output to use in the error message.
+ * @throws Error if the validation fails.
+ */
+function validateAmountIsLesserThanMaxSupply(
+    amount: string,
+    tokenSupply: string,
+    outputName: string
+) {
+    if (bigInt(amount).gt(tokenSupply)) {
+        failValidation(`${outputName} amount field must not be greater than max token supply.`);
+    }
+}
+
+/**
+ * Validate that the unlock conditions define only allowed types.
+ * @param outputType The type of the output.
+ * @param unlockConditions The amount to validate.
+ * @param outputName The name of the output to use in the error message.
+ * @throws Error if the validation fails.
+ */
+function validateUnlockConditionAllowedTypes(
+    outputType: number,
+    unlockConditions: UnlockConditionTypes[],
+    outputName: string
+) {
+    const allowedUnlockConditionTypes = OUTPUT_TYPE_TO_SUPPORTED_PROP_VALUES.get(outputType)?.unlockConditions ?? [];
+
+    if (!unlockConditions.every(uC => allowedUnlockConditionTypes.includes(uC.type))) {
+        const unlockConditionNames = UNLOCK_CONDITION_TYPE_NAMES.filter(
+            (uC, index) => allowedUnlockConditionTypes.includes(index)
```

Comment:
At first i thought that this thing with the index was smelly (`uC` is unused we should `_` it btw), but now i think i see, it's sneaky not smelly.
This relies on the fact that the strings in `UNLOCK_CONDITION_TYPE_NAMES` are **ordered** in the exact ordering like the outputs ordering by `type`, so that when you get the index you can pass it in `includes`, and all matches, am i right ?
If so, it's a nice trick but very sneaky :D
If someone reorders the lines in that innocent looking string array this code starts failing, we should probably make it easier to understand.